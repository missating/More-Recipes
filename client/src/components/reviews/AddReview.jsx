import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

// validations
import reviewValidator from '../../validation/reviewValidator';

// actions
import addReview from '../../actions/addReview';

/**
 * @description Creates AddReview component
 *
 * @class AddReview
 *
 * @extends {React.Component}
 */
export class AddReview extends React.Component {
  /**
   * @description Creates an instance of AddReview.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof AddReview
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      content: '',
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }


  /**
  * @description Bind the value of the inputs to state
  *
  * @method onChange
  *
  * @param {object} event
  *
  * @memberof AddReview
  *
  * @returns {void}
  */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }


  /**
   * @description Submit the form
   *
   * @method onSubmit
   *
   * @param {object} event
   *
   * @memberof AddReview
   *
   * @returns {void}
   */
  onSubmit(event) {
    event.preventDefault();
    const { content } = this.state;
    const { recipeId } = this.props;
    const isValid = this.isValid();
    if (isValid) {
      this.props.addNewReview(content, recipeId)
        .then(() => {
          this.setState({
            content: ''
          });
        });
    }
  }


  /**
    * @description Validates user's data before making post request
    *
    * @method isValid
    *
    * @memberof AddReview
    *
    * @returns {boolean} true or false
    */
  isValid() {
    const { errors, isValid } = reviewValidator(this.state);
    if (!isValid) {
      this.setState({ errors });
    } else {
      this.setState({ errors: {} });
      return isValid;
    }
  }


  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof AddReview
   *
   * @returns {void}
   *
   */
  render() {
    return (
      <div className="container">
        <form onSubmit={this.onSubmit}>
          <textarea
            className="form-control"
            name="content"
            id="reviewContent"
            rows="4"
            cols="30"
            placeholder="Add a review..."
            onChange={this.onChange}
            value={this.state.content}
          />
          {
            this.state.errors.content &&
            <span
              className="help-block text-danger"
            >
              {this.state.errors.content}
            </span>
          }

          <br />
          <button className="btn btn-secondary" style={{ float: 'right' }}>
            SUBMIT
          </button>
        </form>
      </div>
    );
  }
}
AddReview.propTypes = {
  recipeId: PropTypes.number.isRequired,
  addNewReview: PropTypes.func.isRequired,
};


const mapDispatchToProps = dispatch => ({
  addNewReview: (content, recipeId) => dispatch(addReview(content, recipeId))
});

const ConnectedComponent = connect(null, mapDispatchToProps)(AddReview);
export { ConnectedComponent as ConnectedAddReview };

