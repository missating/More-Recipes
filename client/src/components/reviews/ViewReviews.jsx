import React from 'react';
import PropTypes from 'prop-types';
import Gravatar from 'react-gravatar';

const ViewReviews = props => (
  <div className="container">
    <div className="image">
      <Gravatar
        className="img rounded"
        email={props.email}
      />
    </div>
    <p> {props.content} </p>
    <h5> <em> {props.user}</em> </h5>
    <hr />
  </div>
);

ViewReviews.propTypes = {
  content: PropTypes.string.isRequired,
  user: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired
};

export default ViewReviews;
