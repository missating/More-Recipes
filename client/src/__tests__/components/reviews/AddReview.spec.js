import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import { AddReview } from '../../../components/reviews/AddReview';

describe('Add review component', () => {
  const props = {
    recipeId: 1,
    addNewReview: jest.fn(() => Promise.resolve())
  };

  it('Should render', () => {
    const wrapper = shallow(<AddReview
      {...props}
    />);
    expect(wrapper.find('form').exists).toBeTruthy();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it(
    'calls onChange method when the value of the new review input changes',
    () => {
      const spy = sinon.spy(AddReview.prototype, 'onChange');
      const wrapper = shallow(<AddReview
        {...props}
      />);
      wrapper.find('#reviewContent').simulate('change', {
        target: { name: 'content', value: 'good stuff' }
      });
      expect(spy.called).toBeTruthy();
    }
  );

  it('Should submit form when submit button is clicked', () => {
    const spy = sinon.spy(AddReview.prototype, 'onSubmit');
    const wrapper = shallow(<AddReview
      {...props}
    />);
    wrapper.instance().setState({
      content: 'good stuff',
      errors: {}
    });
    wrapper.update();
    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });
    expect(spy.called).toBeTruthy();
    wrapper.instance().setState({
      content: '',
      errors: {}
    });
    wrapper.update();
    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });
    expect(wrapper.instance().state.errors)
      .toEqual({ content: 'Please add a review' });
  });
});

