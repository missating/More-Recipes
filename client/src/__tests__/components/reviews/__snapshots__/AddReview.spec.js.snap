// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add review component Should render 1`] = `
<div
  className="container"
>
  <form
    onSubmit={[Function]}
  >
    <textarea
      className="form-control"
      cols="30"
      id="reviewContent"
      name="content"
      onChange={[Function]}
      placeholder="Add a review..."
      rows="4"
      value=""
    />
    <br />
    <button
      className="btn btn-secondary"
      style={
        Object {
          "float": "right",
        }
      }
    >
      SUBMIT
    </button>
  </form>
</div>
`;
